### PR TITLE
Gh 9374 from 6.0

### PR DESCRIPTION
**Pull request by Rossh87**
**Summary**
Implements sensible defaults discussed in #9374 if connection readPreference is 'secondary'  or 'preferSecondary'.  If readPreference will be set to 'nearest' or 'primaryPreferred', it warns but otherwise behaves as before.
**Examples**
Several tests added for Mongoose.prototype.connect and Mongoose.prototype.createConnection.

### CODE REVIEW

File: lib/helpers/processConnectionOptions.js [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    const configIdx = opts.config && opts.config.autoIndex;
+    const { autoCreate, autoIndex, useCreateIndex } = opts;
+
+    // don't silently override user-provided indexing options
+    if (configIdx || autoCreate || autoIndex || useCreateIndex) {
+        const conflictLevel = getWarningLevel(pref);
+
+        if (conflictLevel === warnLevels.FAIL) {
+            throwReadPreferenceError();
+        } else if (conflictLevel === warnLevels.WARN) {
```
Comment (vkarpov15) [inline lib/helpers/processConnectionOptions.js:27]: I'd recommend getting rid of the `WARN` logic entirely, unless we add an extra option to explicitly suppress the warning.

File: test/collection.test.js [hunk 2]
```diff
@@ -7,140 +7,140 @@ const assert = require('assert');
...
-        db.close();
-        done();
-      });
-    }
-
-    collection.insertOne({ foo: 'bar' }, {}, function(err, result) {
-      assert.ok(connected);
-      insertedId = result.insertedId;
-      finish();
+describe('collections:', function () {
```
Comment (vkarpov15) [inline test/collection.test.js:10]: It looks like some linter or formatter made a lot of unnecessary changes in this. Can you please undo these unnecessary changes and just commit the tests you're adding without the extra formatting diff?

File: lib/helpers/processConnectionOptions.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+        opts.config = { autoIndex: false };
+        opts.autoCreate = false;
+        opts.autoIndex = false;
+        opts.useCreateIndex = false;
+        return opts;
+    }
+}
+
+function getUriReadPreference(connectionString) {
+    const exp = /(?:&|\?)readPreference=(\w+)(?:&|$)/;
```
Comment (vkarpov15) [inline lib/helpers/processConnectionOptions.js:47]: Nice :+1:

**Review by vkarpov15 — CHANGES_REQUESTED** [review]

File: lib/helpers/processConnectionOptions.js (hunk 1 above)
Comment (Rossh87) [reply lib/helpers/processConnectionOptions.js:27]: No problem.  It seems like we could eliminate the warning behavior in 1 of 4 ways:
1.  Permit the user to set index-creation options for connections that *could* connect to a non-primary and fail somewhat cryptically.
2. Throw for `nearest` and `primaryPreferred` in the same way as `secondary` and `secondaryPreferred`.
3. Override the user's preference and default to false anyway.

I *think* I prefer 2, but after some consideration I'm not sure that silently overriding is as bad as I first thought.  Maybe a user who's savvy enough to notice the override is also savvy enough not to set indexing options on a secondary connection, so we'd end up only throwing for less-familiar users who might prefer a connection that Just Works(TM).  In that case we could cut out the error path entirely, which would simplify the code a fair bit.  Do you have a preference?

File: test/collection.test.js (hunk 2 above)
Comment (Rossh87) [reply test/collection.test.js:10]: Yep, sorry, should have run the linter before commit.

File: lib/helpers/processConnectionOptions.js (hunk 1 above)
Comment (vkarpov15) [reply lib/helpers/processConnectionOptions.js:27]: I prefer (1) actually. Index builds only happen once when the app starts up, so it is reasonable to connect using `primaryPreferred` for durability with the assumption that, when the app starts, the primary will be fine, and you're ok with the index build failing if there isn't a primary.

Comment (Rossh87) [reply lib/helpers/processConnectionOptions.js:27]: Cool, sounds good.  I've got all these changes ready to go.  It looks like 6.0 has moved on since I forked it--would you like me to merge/rebase those changes into this PR, or just fix the commented issues?  Thanks!

Comment (Rossh87) [reply lib/helpers/processConnectionOptions.js]: I went ahead and implemented all the changes from this PR/comments against the latest commit of 6.0, and rolled the changes up into a single commit.  The warn logic and all the unnecessary formatting diff should be gone.  The tests that are failing on this commit are failing on the tip commit of 6.0, so I don't think I introduced any new issues.  Please let me know if there's anything else you'd like changed.  Thanks!

**Review by vkarpov15 — APPROVED** [review]
Nice, thanks :+1:
